### PR TITLE
Issue #1192 Separate local persistence from file based persistence

### PR DESCRIPTION
**Pull request by rkavanap**
Refactor to separate local persistence service from file based persistence.

### CODE REVIEW

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 1]
```diff
@@ -17,22 +17,71 @@
...
+   * @return Identifier that can be used to identify the safe space.
+   */
+  String prepareSafeSpace(String owner, String name);
+
+  /**
+   * Create the safe spaces in actual physical space for the owner.
+   *
+   * @param safeSpaceId prepared logical space on which the physical space
+   */
+  void createSafeSpace(String safeSpaceId) throws CachePersistenceException;
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: Please, let's not use String there but instead some opaque id with a specific type.

File: impl/src/main/java/org/ehcache/impl/internal/store/disk/OffHeapDiskStore.java [hunk 2]
```diff
@@ -327,6 +328,9 @@ public int rankAuthority(ResourceType<?> authorityResource, Collection<ServiceCo
       if (serviceProvider == null) {
         throw new NullPointerException("ServiceProvider is null in OffHeapDiskStore.Provider.");
       }
+      if (diskPersistenceService == null) {
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/disk/OffHeapDiskStore.java]: That error here makes no sense - do the check in `start` instead

File: impl/src/main/java/org/ehcache/impl/persistence/DefaultFileBasedPersistenceService.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Default implementation of the {@link FileBasedPersistenceService} which can be used explicitly when
+ * {@link org.ehcache.PersistentUserManagedCache persistent user managed caches} are desired.
+ *
+ * @author RKAV refactored for issue-1192
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultFileBasedPersistenceService.java]: Drop `@author` tag - we are not using them

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 4]
```diff
@@ -17,22 +17,71 @@
...
+
+  /**
+   * Prepare a logical safe directory space for the owner and return an identifying space Id.
+   *
+   * @param owner Service owner that owns the safe space.
+   * @param name Identifying name for the space.
+   *
+   * @return Identifier that can be used to identify the safe space.
+   */
+  String prepareSafeSpace(String owner, String name);
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: What exactly is the `owner` concept providing?

File: impl/src/main/java/org/ehcache/impl/persistence/DefaultFileBasedPersistenceService.java [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+  public FileBasedPersistenceContext createPersistenceContextWithin(PersistenceSpaceIdentifier<?> identifier, String name)
+      throws CachePersistenceException {
+    if (containsSpace(identifier)) {
+      return persistenceService.createContextWithin(identifier.toString(), name);
+    } else {
+      throw new CachePersistenceException("Unknown space: " + identifier);
+    }
+  }
+
+  //////PRIVATE/////////
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultFileBasedPersistenceService.java]: Please match code style - we do not use such separators

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java (hunk 4 above)
Comment (rkavanap) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: In case there are both frs and disk (and maybe hybrid) under the same cache manager. So when I destroy one it does not destroy the other..

**Comment by ljacomet** [comment]
Given the release of 3.1.0, this PR cannot be merged before master has been moved to 3.2.0-SNAPSHOT

File: impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java [hunk 6]
```diff
@@ -16,74 +16,44 @@
...
-    ILLEGALS.add('"');
-    ILLEGALS.add('|');
-    ILLEGALS.add('?');
-    ILLEGALS.add('*');
-    ILLEGALS.add('.');
-  }
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultLocalPersistenceService.class);
 
-  private final ConcurrentMap<String, PersistenceSpace> knownPersistenceSpaces = new ConcurrentHashMap<String, PersistenceSpace>();
+  private final ConcurrentMap<String, SafeSpace> createdSpaces = new ConcurrentHashMap<String, SafeSpace>();
```
Comment (albinsuresh) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java]: Is it really required to maintain this map internally? The user of this service should be maintaining the reference to the `SafeSpaceIdentifier` instances that it prepares? Isn't the `SafeSpaceIdentifier` just a logical representation of the `SafeSpace`? And the `SafeSpaceIdentifier` generation is idempotent using the same owner and identifier values, right? So I'm not sure if we need to maintain it here. Let the caller keep the references to the `SafeSpaceIdentifier` that it has prepared or prepare a new one which will provide the same logical `SafeSpaceIdentifier`. I'm viewing this `SafeSpaceIdentifier` as analogous to `File` which is just a logical representation of a physical file just like `SafeSpaceIdentifier` is for a `SafeSpace`. May be there is something that I'm missing.

File: impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java [hunk 7]
```diff
@@ -153,363 +101,196 @@ public synchronized void stop() {
...
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
-  public boolean handlesResourceType(ResourceType<?> resourceType) {
-    return ResourceType.Core.DISK.equals(resourceType);
-  }
+  public SafeSpaceIdentifier prepareSafeSpace(String owner, String identifier) {
```
Comment (albinsuresh) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java]: The name as _prepareSafeSpaceIdentifier_ or __createSafeSpaceIdentifier_, may be? Since it is just about creating the identifier? Just a personal opinion.

File: impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java [hunk 8]
```diff
@@ -153,363 +97,168 @@ public synchronized void stop() {
...
-  private File getDirectoryFor(String identifier) {
-    File directory = new File(rootDirectory, safeIdentifier(identifier));
-
-    for (File parent = directory.getParentFile(); parent != null; parent = parent.getParentFile()) {
-      if (rootDirectory.equals(parent)) {
-        return directory;
-      }
+  @Override
+  public synchronized void startForMaintenance(ServiceProvider<MaintainableService> serviceProvider) {
+    if (rootDirectory == null) {
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java]: Why is that not an error?

File: impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java [hunk 9]
```diff
@@ -153,363 +97,168 @@ public synchronized void stop() {
...
-  }
 
-  private static MessageDigest getSha1Digest() {
-    try {
-      return MessageDigest.getInstance("SHA-1");
-    } catch (NoSuchAlgorithmException e) {
-      throw new AssertionError("All JDKs must have SHA-1");
-    }
+  private SafeSpace createSafeSpaceLogical(String owner, String identifier) {
+    File ownerDirectory = new File(rootDirectory, owner);
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java]: Is owner considered _safe_ because it is provided by Ehcache services and not user input?

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 10]
```diff
@@ -17,22 +17,64 @@
...
+
+  /**
+   * Create a physical persistent context (directory) within the safe space.
+   *
+   * @param safeSpaceId Safe space identifier
+   * @return persistent context
    */
-  FileBasedPersistenceContext createPersistenceContextWithin(PersistenceSpaceIdentifier<?> identifier, String name) throws CachePersistenceException;
-}
+  FileBasedPersistenceContext createContextWithin(SafeSpaceIdentifier safeSpaceId, String name) throws CachePersistenceException;
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: As discussed with @rkavanap this method does not belong here

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 11]
```diff
@@ -17,22 +17,64 @@
...
+  void destroyAll(String owner);
+
+  /**
+   * Destroy safe space given owner and identifier.
+   *
+   * @param owner owner of safe spaces
+   * @param name name of safe spaces
+   * @param verbose log more
+   */
+  void destroySafeSpace(String owner, String name, boolean verbose);
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: Given that the `SafeSpaceIdentifier` can be obtained without any file based action, this makes little sense as one can always invoke the overloaded method.

**Comment by ljacomet** [comment]
Still having issues with the naming here.

`LocalPersistenceService` and `FileBasedPersistenceService` are names too close to each other.

The first one represents the ability to obtain a `FileBasedPersistenceContext` to store information on disk. The second one is tailored for the needs of `DISK` resource and a close companion to the `DiskStore.Provider`.

Should they be named: `FileBasedPersistenceService` and DiskResourceService`? any other proposals?

File: impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java (hunk 9 above)
Comment (rkavanap) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java]: Yes. owner is typically hardcoded, pre-defined values used by Ehcache persistable resource services only. It is to identify a root for a given service that is separate from other services,

File: impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java (hunk 8 above)
Comment (rkavanap) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java]: yes. this should be an error. Infact, this check is not even necessary as this object will not exist if rootDirectory is null...

**Comment by rkavanap** [comment]
Addressed @ljacomet comments into 2 separate commits. 1 commit has the renames and 1 commit has other changes Will squash into a single commit post final review. 
 
Btw, I have not renamed LocalPersistenceService, since FileBasedPersistenceService is now renamed to DiskResourceService that reduced the confusion that @ljacomet stated above. But if @ljacomet  still thinks LocalPeristenceService should be renamed, it can be done in another commit (which can also then rename FileBasedStateRepository to DiskResourceStateRepository)..

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 12]
```diff
@@ -17,22 +17,50 @@
...
 
   /**
-   * Creates a new persistence context within the given space.
+   * Create the safe spaces in actual physical space for the owner.
    *
-   * @param identifier space to create within
-   * @param name name of the context to create
-   * @return a {@link FileBasedPersistenceContext}
+   * @param safeSpaceId prepared logical space on which the physical space
    */
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: missing javadoc on `name` param

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 13]
```diff
@@ -17,22 +17,50 @@
...
+   * @param owner Service owner that owns the safe space.
+   * @param name Identifying name for the space.
+   *
+   * @return Opaque Identifier that can be used to identify the safe space.
+   */
+  SafeSpaceIdentifier createSafeSpaceIdentifier(String owner, String name);
 
   /**
-   * Creates a new persistence context within the given space.
+   * Create the safe spaces in actual physical space for the owner.
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: Typo: Creates

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 14]
```diff
@@ -17,22 +17,50 @@
...
+  void destroySafeSpace(SafeSpaceIdentifier safeSpaceId, boolean verbose);
+
+  /**
+   * Destroy all safe spaces provided to this owner.
+   *
+   * @param owner owner of safe spaces.
+   */
+  void destroyAll(String owner);
+
+  // identifier that also returns the root of the space
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: Let's have proper javadoc here on the interface and its method. Especially as there is a catch on the `File` return, you cannot assume it exists ... and you should not be the one creating it ...

File: impl/src/main/java/org/ehcache/config/builders/UserManagedCacheBuilder.java [hunk 15]
```diff
@@ -207,7 +207,7 @@ T build(ServiceLocator serviceLocator) throws IllegalStateException {
       if (id == null) {
         throw new IllegalStateException("Persistent user managed caches must have an id set");
       }
-      LocalPersistenceService persistenceService = serviceLocator.getService(LocalPersistenceService.class);
+      DiskResourceService persistenceService = serviceLocator.getService(DiskResourceService.class);
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/config/builders/UserManagedCacheBuilder.java]: lets rename the variable as well.

File: impl/src/main/java/org/ehcache/config/builders/UserManagedCacheBuilder.java [hunk 16]
```diff
@@ -296,8 +296,8 @@ public void close() throws Exception {
     eventDispatcher.setStoreEventSource(store.getStoreEventSource());
 
     if (persistent) {
-      LocalPersistenceService persistenceService = serviceLocator
-          .getService(LocalPersistenceService.class);
+      DiskResourceService persistenceService = serviceLocator
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/config/builders/UserManagedCacheBuilder.java]: lets rename the variable as well

File: impl/src/main/java/org/ehcache/impl/internal/persistence/DefaultFileBasedPersistenceServiceFactory.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.ehcache.core.spi.service.DiskResourceService;
+import org.ehcache.core.spi.service.ServiceFactory;
+import org.ehcache.impl.persistence.DefaultDiskResourceService;
+import org.ehcache.spi.service.ServiceCreationConfiguration;
+
+/**
+ *
+ */
+public class DefaultFileBasedPersistenceServiceFactory implements ServiceFactory<DiskResourceService> {
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/persistence/DefaultFileBasedPersistenceServiceFactory.java]: That needs a rename to match the renamed service

File: core/src/main/java/org/ehcache/core/spi/service/DiskResourceService.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.core.spi.service;
+
+import org.ehcache.CachePersistenceException;
+import org.ehcache.spi.persistence.PersistableResourceService;
+
+/**
+ * Service to provide file based persistence context to caches tied to disk based stores.
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/DiskResourceService.java]: I would rephrase this class level doc (in place of the two paragraphs you have):

> Resource service handling file level operations for disk tiers.

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 19]
```diff
@@ -17,22 +17,63 @@
...
+import org.ehcache.spi.service.MaintainableService;
+
+import java.io.File;
 
 /**
- * Service to provide persistence context to caches requiring it.
- * <P>
- * Will be used by caches with a disk store, whether or not the data should survive a program restart.
- * <P/>
+ * Service that provides unique local persistence spaces (and safe sub spaces) to any service that requires it
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: > unique local persistence spaces (and safe sub spaces)

changed to
> isolated persistence spaces

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 20]
```diff
@@ -17,22 +17,63 @@
...
    *
-   * @param identifier space to create within
-   * @param name name of the context to create
-   * @return a {@link FileBasedPersistenceContext}
+   * @return Opaque Identifier that can be used to identify the safe space.
+   */
+  SafeSpaceIdentifier createSafeSpaceIdentifier(String owner, String name);
+
+  /**
+   * Creates or finds the safe space represented by {@code safeSpaceId} in the underlying physical space.
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: > Creates or finds ...

changed to
> Creates if it does not exist ...

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java [hunk 21]
```diff
@@ -17,22 +17,63 @@
...
+   * @param owner owner of safe spaces.
+   */
+  void destroyAll(String owner);
+
+  /**
+   * Identifier to the logical safe space
    */
-  FileBasedPersistenceContext createPersistenceContextWithin(PersistenceSpaceIdentifier<?> identifier, String name) throws CachePersistenceException;
-}
+  interface SafeSpaceIdentifier {
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: Should we rename that do `FileSpaceIdentifier`? Not sure the `Safe` brings anything.

File: impl/src/main/java/org/ehcache/impl/internal/persistence/DefaultDiskResourceServiceFactory.java [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package org.ehcache.impl.internal.persistence;
+
+import org.ehcache.core.spi.service.DiskResourceService;
+import org.ehcache.core.spi.service.ServiceFactory;
+import org.ehcache.impl.persistence.DefaultDiskResourceService;
+import org.ehcache.spi.service.ServiceCreationConfiguration;
+
+/**
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/persistence/DefaultDiskResourceServiceFactory.java]: Get rid of the empty javadoc block

File: core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java (hunk 21 above)
Comment (rkavanap) [inline core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java]: Isn't SafeSpaceIdentifier consistent with the rest of the calls (such as createSafeSpaceId..etc). So should I rename every method in the interface to FileSpace from SafeSpace?.

Isn't FileSpace pleonastic? or another way to look at it is space for file (which is directory). Since the service does a bit more than providing a directory (such as locking to prevent accidental use of the same directory under different cache managers and validating and unique naming etc.) thought SafeSpace conveys the meaning better.

**Comment by rkavanap** [comment]
test this please

File: impl/src/main/java/org/ehcache/impl/persistence/DefaultDiskResourceService.java [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+      return null;
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void destroy(String name) throws CachePersistenceException {
+    if (persistenceService == null) {
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultDiskResourceService.java]: So because it is null, you have no option to start it even if you wanted to. And that explains the ignored tests.
Let me think about this ...

Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultDiskResourceService.java]: Let's go ahead with this merge but not close #1192 under which I want to address re-enabling the tests.
So go ahead, rebase and squash commits.

Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/persistence/DefaultDiskResourceService.java]: @henri-tremblay May want to give it some thought. Because this is now decomposed in two services, the solution we came up with for 3.1.2 no longer works. However, we should stick with the public behaviour IMO. Just find a different implementation to fix the problem.
